### PR TITLE
Config to json

### PR DESCRIPTION
**Pull request by VicDeo**
**Contains a migration - needs minor version bump for the app**

Make transition to new Wizard easier by
- [x] migrating to json-encoded server config in advance.
- [x] Translate server configuration from underscore to pascalCase only when dealing with frontend (before this PR it was done for the database too)

Before:
```
select configkey, configvalue from oc_appconfig where appid='user_ldap' and configkey  like 's01%';
+--------------------------------------+----------------+
| configkey                            | configvalue    |
+--------------------------------------+----------------+
| s01has_memberof_filter_support       | 0              |
| s01home_folder_naming_rule           |                |
| s01last_jpegPhoto_lookup             | 0              |
| s01ldap_agent_password               |                |
| s01ldap_attributes_for_group_search  |                |
| s01ldap_attributes_for_user_search   |                |
| s01ldap_backup_host                  |                |
| s01ldap_backup_port                  |                |
| s01ldap_base                         |                |
| s01ldap_base_groups                  |                |
| s01ldap_base_users                   |                |
| s01ldap_cache_ttl                    | 600            |
| s01ldap_configuration_active         | 0              |
| s01ldap_display_name                 | displayName    |
| s01ldap_dn                           |                |
| s01ldap_dynamic_group_member_url     |                |
| s01ldap_email_attr                   |                |
| s01ldap_experienced_admin            | 0              |
| s01ldap_expert_username_attr         |                |
| s01ldap_expert_uuid_group_attr       |                |
| s01ldap_expert_uuid_user_attr        |                |
| s01ldap_group_display_name           | cn             |
| s01ldap_group_filter                 |                |
| s01ldap_group_filter_mode            | 0              |
| s01ldap_group_member_assoc_attribute | uniqueMember   |
| s01ldap_groupfilter_groups           |                |
| s01ldap_groupfilter_objectclass      |                |
| s01ldap_host                         |                |
| s01ldap_login_filter                 |                |
| s01ldap_login_filter_mode            | 0              |
| s01ldap_loginfilter_attributes       |                |
| s01ldap_loginfilter_email            | 0              |
| s01ldap_loginfilter_username         | 1              |
| s01ldap_nested_groups                | 0              |
| s01ldap_network_timeout              | 2              |
| s01ldap_override_main_server         |                |
| s01ldap_paging_size                  | 500            |
| s01ldap_port                         |                |
| s01ldap_quota_attr                   |                |
| s01ldap_quota_def                    |                |
| s01ldap_tls                          | 0              |
| s01ldap_turn_off_cert_check          | 0              |
| s01ldap_user_display_name_2          |                |
| s01ldap_user_filter_mode             | 0              |
| s01ldap_user_name                    | samaccountname |
| s01ldap_userfilter_groups            |                |
| s01ldap_userfilter_objectclass       |                |
| s01ldap_userlist_filter              |                |
| s01use_memberof_to_detect_membership | 1              |
+--------------------------------------+----------------+
```

After
```
select configkey, configvalue from oc_appconfig where appid='user_ldap' and configkey  = 'conn-s01';
| conn-s01  | {"hasMemberOfFilterSupport":"0","homeFolderNamingRule":"","lastJpegPhotoLookup":"0","ldapAgentPassword":"","ldapAttributesForGroupSearch":"","ldapAttributesForUserSearch":"","ldapBackupHost":"","ldapBackupPort":"","ldapBase":"","ldapBaseGroups":"","ldapBaseUsers":"","ldapCacheTTL":"600","ldapConfigurationActive":"0","ldapUserDisplayName":"displayName","ldapAgentName":"","ldapDynamicGroupMemberURL":"","ldapEmailAttribute":"","ldapExperiencedAdmin":"0","ldapExpertUsernameAttr":"","ldapExpertUUIDGroupAttr":"","ldapExpertUUIDUserAttr":"","ldapGroupDisplayName":"cn","ldapGroupFilter":"","ldapGroupFilterMode":"0","ldapGroupMemberAssocAttr":"uniqueMember","ldapGroupFilterGroups":"","ldapGroupFilterObjectclass":"","ldapHost":"","ldapLoginFilter":"","ldapLoginFilterMode":"0","ldapLoginFilterAttributes":"","ldapLoginFilterEmail":"0","ldapLoginFilterUsername":"1","ldapNestedGroups":"0","ldapNetworkTimeout":"2","ldapOverrideMainServer":"","ldapPagingSize":"500","ldapPort":"","ldapQuotaAttribute":"","ldapQuotaDefault":"","ldapTLS":"0","turnOffCertCheck":"0","ldapUserDisplayName2":"","ldapUserFilterMode":"0","ldapUserName":"samaccountname","ldapUserFilterGroups":"","ldapUserFilterObjectclass":"","ldapUserFilter":"","useMemberOfToDetectMembership":"1"} |
```

Note: I tried to made the migration as independent as possible so the changes in app won't break it in future.

### CODE REVIEW

**Comment by VicDeo** [comment]
Note: acceptance tests are failing randomly
It looks like that success of `user_ldap/tests/acceptance/features/webUIProvisioning/groups.feature:18` depends on `/dev/urandom`

File: appinfo/Migrations/Version20190315192717.php [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+			\json_encode($legacyConfig)
+		);
+	}
+
+	/**
+	 * Deletes a given legacy LDAP/AD server configuration.
+	 *
+	 * @param string $prefix the configuration prefix of the config to delete
+	 *
+	 * @return bool true on success, false otherwise
```
Comment (jvillafanez) [inline appinfo/Migrations/Version20190315192717.php]: It never returns false...

File: appinfo/Migrations/Version20190315192717.php [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+	/**
+	 * Get all possible configkey names prefixed with a given string
+	 *
+	 * @param string $prefix
+	 *
+	 * @return array
+	 */
+	protected function getPrefixedLegacyKeys($prefix) {
+		$keys = \array_keys($this->knownLegacyKeys);
```
Comment (jvillafanez) [inline appinfo/Migrations/Version20190315192717.php]: it feels easier to read a classic foreach loop:
```
$prefixed = [];
foreach ($this->knownLegacyKeys as $key => $value) {
  $prefixedKey = "{$prefix}{$key}";
  $prefixed[] = $prefixedKey;
}
```

**Comment by jvillafanez** [comment]
The code for the migration seems quite heavy, but since it will be executed only one time I guess we can overlook some performance problems

File: appinfo/Migrations/Version20190315192717.php [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+			)
+			->setParameter('keys', $this->getPrefixedLegacyKeys($prefix), IQueryBuilder::PARAM_STR_ARRAY);
+
+		$result = $qb->execute();
+		$configValues = $result->fetchAll();
+
+		$current = [];
+		foreach ($configValues as $row) {
+			// if there is a prefix we need to remove its first occurrence in key
+			$key = $prefix === ''
```
[the remark this answers is not in the document]
Comment (VicDeo) [reply appinfo/Migrations/Version20190315192717.php]: ternary operation is a classic too. See Kernighan and Ritchie "The C Programming Language", 1978  2.11 Conditional Expressions :upside_down_face:

File: appinfo/Migrations/Version20190315192717.php [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+				? $row['configkey']
+				: \implode('', \explode($prefix, $row['configkey'], 2));
+			$translatedKey = $this->translations[$key];
+			$current[$translatedKey] = $row['configvalue'];
+		}
+		\ksort($this->translations);
+		\ksort($this->knownLegacyKeys);
+		$defaults = \array_combine(
```
[the remark this answers is not in the document]
Comment (VicDeo) [reply appinfo/Migrations/Version20190315192717.php]: I don't think it's two complex - 3 lines only (as long as merging defaults with non-defaults is a common practice). Personally I don't like researching what we are taking from **this** array to use it as a key in **that** array to produce another **that** array in the end.

File: appinfo/Migrations/Version20190315192717.php (hunk 2 above)
Comment (VicDeo) [reply appinfo/Migrations/Version20190315192717.php]: ...for those who never read about array functions. :)

File: appinfo/Migrations/Version20190315192717.php [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+	}
+
+	/**
+	 * Deletes a given legacy LDAP/AD server configuration.
+	 *
+	 * @param string $prefix the configuration prefix of the config to delete
+	 *
+	 * @return bool true on success, false otherwise
+	 */
+	public function deleteLegacyConfig($prefix) {
```
[the remark this answers is not in the document]
Comment (VicDeo) [reply appinfo/Migrations/Version20190315192717.php]: done

File: appinfo/Migrations/Version20190315192717.php [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+		return \array_merge($defaults, $current);
+	}
+
+	/**
+	 * Store config as json
+	 *
+	 * @param string $prefix
+	 * @param array $legacyConfig
+	 */
+	protected function convertConfig($prefix, $legacyConfig) {
```
[the remark this answers is not in the document]
Comment (VicDeo) [reply appinfo/Migrations/Version20190315192717.php]: renamed to `storeConfig` and `$configData`

File: appinfo/Migrations/Version20190315192717.php (hunk 2 above)
Comment (VicDeo) [reply appinfo/Migrations/Version20190315192717.php]: changed nevertheless

File: appinfo/Migrations/Version20190315192717.php (hunk 1 above)
Comment (VicDeo) [reply appinfo/Migrations/Version20190315192717.php]: true. I forgot to update the signature.

File: appinfo/Migrations/Version20190315192717.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (VicDeo) [reply appinfo/Migrations/Version20190315192717.php]: changed to 
```
		foreach ($this->translations as $legacyKey => $newKey) {
			if (!isset($current[$newKey])) {
				$current[$newKey] = $this->knownLegacyKeys[$legacyKey];
			}
		}
```

**Comment by VicDeo** [comment]
@jvillafanez everything except ternary operation is addressed.

**Comment by PVince81** [comment]
@cdamken @pako81 FYI, this would mean that now one cannot easily change individual values as before, not sure if this is a concern. this is required by the new ldap wizard

**Comment by pako81** [comment]
> @cdamken @pako81 FYI, this would mean that now one cannot easily change individual values as before, not sure if this is a concern. this is required by the new ldap wizard

You mean directly in config.php? Should not be a big issue AFAICT, Customers usually do this over the LDAP wizard

**Comment by PVince81** [comment]
no, I mean directly in oc_appconfig table.

before this PR change every key had a separate entry

**Comment by pako81** [comment]
Ah ok. No, not an issue at all.

**Pull request merged**